packages/kit/src/core/sync/write_types/index.js

```diff
@@ -188,10 +188,14 @@ function update_types(config, routes, route, to_delete = new Set()) {
 	// add 'Expand' helper
 	// Makes sure a type is "repackaged" and therefore more readable
 	declarations.push('type Expand<T> = T extends infer O ? { [K in keyof O]: O[K] } : never;');
+
+	// returns the predicate of a matcher's type guard - or string if there is no type guard
 	declarations.push(
-		`type RouteParams = { ${route.params
-			.map((param) => `${param.name}${param.optional ? '?' : ''}: string`)
-			.join('; ')} }`
+		'//@ts-ignore\ntype MatcherParam<M> = M extends (param: string) => param is infer U ? U : string;'
```

Comment:
Both questions kinda have an answer stemming from the following behaviour:

Typescript will complain if a predicate is not assignable to the type of the parameter. For example:
```ts
const matcher = (param: string) : param is 5 => //...
```
would error since `typeof 5` is not assignable to type `string`. 

Since SvelteKit always types the parameter of a matcher as a string it shouldn't be possible to write a non-string predicate on a matcher.

As for the `@ts-ignore`, try using the `MatcherParam` type without it: ([TS Playground Link](https://www.typescriptlang.org/play?ts=5.2.2#code/C4TwDgpgBAsghsAxgCwApwE5wLYGcA8MAfFALyxQQAewEAdgCa5QAUYmOUAXFLsBgEs6AcwCUZEuyzYoA5kIBmEDFACqUAPxruvfkOEBuAFBA))
```ts
type MatcherParam<M> = M extends (param: string) => param is infer U ? U : string;
```
Typescript believes that the type of `infer U` could be anything and therefore isn't assignable to type `string`. This is clearly wrong. The `infer U` should have been type-narrowed to `string` because `param` is typed as string. The predicate checking I described above prevents any other case.

If you really don't like the `@ts-ignore`, we could change the type to:
```ts
type MatcherParam<M> = M extends (param: any /* <- */) => param is infer U ? U : string;
```
This would still work, but I prefer specifying that param is a string, since the fallback (the ` : string` at the end) is a string.

Reply:
I'll defer to @dummdidumm on the question of whether we should have an ignore, but if we want one it sounds like we should use `@ts-expect-error`. We use `@ts-ignore` to signal a to-do that needs to be cleaned up and `@ts-expect-error` where we plan to leave it in that state permanently

Reply:
Makes sense. I changed it to `@ts-expect-error`. Thanks for taking the time!

Reply:
Can we make the type 
```ts
type MatchParams<M> = M extends (param : string) => param is infer U ? U extends string ? U : string : string;
```
That way if someone writes a matcher doing `param is number` (which is wrong, but they might anyway) the type is still `string` and not switched to `number`